core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java

```diff
@@ -16,25 +16,103 @@
+ * <code>SecurityContextHolder</code>.
+ *
+ * <p>
+ * For a detailed background on what this filter is designed to process, refer to
+ * <a href="https://tools.ietf.org/html/rfc6749>RFC 6749, Section 2.3</a>.
+ *
+ * <p>
+ * In summary, this filter is responsible for processing any request that has a HTTP
+ * request header of <code>Authorization</code> with an authentication scheme of
+ * <code>Basic</code> and a Base64-encoded <code>client_id:client_secret</code> token. For
+ * example, to authenticate client "Spring" with client secret "Secret" the following
+ * header would be presented:
+ *
+ * <pre>
+ *
+ *     Authorization: Basic U3ByaW5nOlNlY3JldA==
+ * </pre>
+ *
+ * <p>
+ * If authentication is successful, the resulting {@link Authentication} object will be
+ * placed into the <code>SecurityContextHolder</code>.
+ *
+ * <p>
+ * If authentication fails, the <code>SecurityContextHolder</code> will be cleared.
+ *
  * @author Joe Grandja
+ * @author Patryk Kostrzewa
  */
 public class OAuth2ClientAuthenticationFilter extends OncePerRequestFilter {
-	private AuthenticationManager authenticationManager;
+
+	private static final String TOKEN_ENDPOINT = "/oauth2/token";
+	private final AuthenticationManager authenticationManager;
+	private final ClientAuthenticationConverter authenticationConverter = new ClientAuthenticationConverter();
+
+	/**
+	 * Creates an instance which will authenticate against the supplied
+	 * {@code AuthenticationManager} and which will ignore failed authentication attempts,
+	 * allowing the request to proceed down the filter chain.
+	 *
+	 * @param authenticationManager
+	 * 		the bean to submit authentication requests to
+	 */
+	public OAuth2ClientAuthenticationFilter(AuthenticationManager authenticationManager) {
+		Assert.notNull(authenticationManager, "authenticationManager cannot be null");
+		this.authenticationManager = authenticationManager;
+	}
 
 	@Override
-	protected void doFilterInternal(HttpServletRequest request,
-			HttpServletResponse response, FilterChain filterChain)
+	public void afterPropertiesSet() {
+		Assert.notNull(authenticationManager, "An AuthenticationManager is required");
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
 			throws ServletException, IOException {
 
+		if (TOKEN_ENDPOINT.equals(request.getRequestURI())) {
+			final OAuth2ClientAuthenticationToken authentication = authenticationConverter.convert(request);
+			if (authentication == null) {
+				filterChain.doFilter(request, response);
+				return;
+			}
+			try {
+				final Authentication result = authenticationManager.authenticate(authentication);
+				onSuccessfulAuthentication(request, response, result);
+			} catch (AuthenticationException failed) {
+				onUnsuccessfulAuthentication(request, response, failed);
+				filterChain.doFilter(request, response);
```

Comment:
Failed authentication should not proceed through the chain. The failure handler should respond with error response.